nucliadb_node/src/shards/shard_writer.rs

```diff
@@ -420,59 +502,106 @@ impl ShardWriter {
     pub fn collect_garbage(&self) -> NodeResult<GarbageCollectorStatus> {
         let _lock = self.gc_lock.blocking_lock();
-        let result = write_rw_lock(&self.indexes).vector_writer.garbage_collection();
-        match result {
-            Ok(()) => Ok(GarbageCollectorStatus::GarbageCollected),
-            Err(error) => match error.downcast_ref::<VectorErr>() {
-                Some(VectorErr::WorkDelayed) => Ok(GarbageCollectorStatus::TryLater),
-                _ => Err(error),
-            },
+        let indexes: &mut ShardWriterIndexes = &mut write_rw_lock(&self.indexes);
+
+        let mut gc_results = Vec::with_capacity(indexes.vectors_indexes.len());
+        for (_, vector_writer) in indexes.vectors_indexes.iter_mut() {
+            let result = vector_writer.garbage_collection();
+            gc_results.push(result);
         }
+
+        for result in gc_results {
+            if let Err(error) = result {
+                return match error.downcast_ref::<VectorErr>() {
+                    Some(VectorErr::WorkDelayed) => Ok(GarbageCollectorStatus::TryLater),
+                    _ => Err(error),
+                };
+            }
+        }
+        Ok(GarbageCollectorStatus::GarbageCollected)
     }
 
     #[tracing::instrument(skip_all)]
     pub fn force_garbage_collection(&self) -> NodeResult<GarbageCollectorStatus> {
         let _lock = self.gc_lock.blocking_lock();
-        let result = write_rw_lock(&self.indexes).vector_writer.force_garbage_collection();
-        match result {
-            Ok(()) => Ok(GarbageCollectorStatus::GarbageCollected),
-            Err(error) => match error.downcast_ref::<VectorErr>() {
-                Some(VectorErr::WorkDelayed) => Ok(GarbageCollectorStatus::TryLater),
-                _ => Err(error),
-            },
+        let indexes: &mut ShardWriterIndexes = &mut write_rw_lock(&self.indexes);
+
+        let mut gc_results = Vec::with_capacity(indexes.vectors_indexes.len());
+        for (_, vector_writer) in indexes.vectors_indexes.iter_mut() {
+            let result = vector_writer.force_garbage_collection();
+            gc_results.push(result);
         }
+
+        for result in gc_results {
+            if let Err(error) = result {
+                return match error.downcast_ref::<VectorErr>() {
+                    Some(VectorErr::WorkDelayed) => Ok(GarbageCollectorStatus::TryLater),
+                    _ => Err(error),
+                };
+            }
+        }
+        Ok(GarbageCollectorStatus::GarbageCollected)
     }
 
     #[tracing::instrument(skip_all)]
     pub fn merge(&self, context: MergeContext) -> NodeResult<MergeMetrics> {
-        let runner = read_rw_lock(&self.indexes).vector_writer.prepare_merge(context.parameters)?;
-        let Some(mut runner) = runner else {
-            return Ok(MergeMetrics {
-                merged: 0,
-                left: 0,
-            });
+        // TODO: return metrics by vectorset, not only the deafult one
```

Comment:
This is only running a merge on the default index, unless I missed something. I'd change this TODO to indicate that not only it returns default metrics, but also that it's only merging the default index. Or even better, actually merge all indexes.

Reply:
Yes, I know. I didn't know if I wanted to change merge protos too